### DISCUSSION TITLE
How can I detect if I'm running the frontend in a browser?

### DISCUSSION THREAD

**Discussion by mettz** [tauri-apps/tauri#6119] [Q&A] [answered]
Hi, folks! Really enjoying Tauri so far! I've just a small issue. Is there a standard way to detect if the frontend of the app is currently running in a browser or in the Tauri webview?

**Comment by FabianLars** [comment, marked as answer]
You can check if `window.__TAURI__` exists.
```ts
if (window.__TAURI__) {
  // inside tauri
}
```

**Reply by mettz** [reply]
It worked like a charm! Thank you very much!

**Comment by zpng** [comment]
![image](https://github.com/tauri-apps/tauri/assets/8509438/19527ca6-eb72-4d83-9fdd-e568eeebd443)

./src/components/sidebar/index.tsx:50:53
Type error: Property '__TAURI__' does not exist on type 'Window & typeof globalThis'.

  48 | 
  49 |     const handleNavigation = (path: string) => {
> 50 |         if (typeof window !== 'undefined' && window.__TAURI__) {
     |                                                     ^
  51 |             // 使用 Tauri 的跳转方式
  52 |             console.log("in tauri")
  53 |             window.location.href = `tauri://${path}`;
 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed with exit code 1.
       Error beforeBuildCommand `pnpm export` failed with exit code 1


why i meet the compile error?

**Reply by FabianLars** [reply]
because typescript only knows about "well known"/standard window properties at build time. You can tell it about the `__TAURI__` property by adding this to your index file:
```ts
declare global {
    interface Window {
        __TAURI__: Record<string, unknown>;
    }
}
```

**Reply by intfract** [reply]
Alternatively, you can check use the `in` keyword to check whether an object has a property in TypeScript. 
```ts
if ('__TAURI__' in window) {
  // inside tauri
}
```

**Reply by liyasthomas** [reply]
This doesn't seem to be working on Tauri v2.

**Reply by FabianLars** [reply]
Yeah iirc we only set `__TAURI__` now if withGlobalTauri is enabled. Try `__TAURI_INTERNALS__` instead.

**Comment by burgil** [comment]
I created `isTauri.tsx` with mobile detection support for V2:

```typescript
const isTauri = '__TAURI__' in window;
export default isTauri;
export const isMobile = navigator.maxTouchPoints > 0;
export const isDesktop = !isMobile;
export const isTauriMobile = isTauri && isMobile;
export const isTauriDesktop = isTauri && isDesktop;
```

Example Usage:

```typescript
import isTauri, { isMobile, isDesktop, isTauriMobile, isTauriDesktop } from '../path-to/isTauri'
```
```typescript
        <div style={isTauriDesktop ? {
            backgroundColor: 'orange',
        } : {
            overflowX: 'hidden',
        }}>
```
```typescript
<Route path="/" element={isTauri ? <Navigate to="/app" /> : <LandingPage />} />
```
```typescript
            {isTauriDesktop && <TitleBar />}
```

I tested it on android, the web and as a windows desktop app.

# Update

I added support for the web version, here is a more advanced use case that some might want:

```typescript
const isTauri = '__TAURI__' in window;
export default isTauri; // running on a desktop app or a mobile app - but not in the browser
export const isWeb = !isTauri; // running on the browser on either desktop or mobile - but not as a tauri app

export const isMobile = navigator.maxTouchPoints > 0; // running in mobile either in the browser or as a tauri app
export const isDesktop = !isMobile; // running in desktop either in the browser or as a tauri app

export const isTauriMobile = isTauri && isMobile; // running on mobile as a tauri app - but not on the browser
export const isTauriDesktop = isTauri && isDesktop; // running on desktop as a tauri app - but not on the browser

export const isWebMobile = isWeb && isMobile; // running on mobile in the browser - but not as a tauri app
export const isWebDesktop = isWeb && isDesktop; // running on desktop in the browser - but not as a tauri app
```

**Reply by Muhammed-Rahif** [reply]
Update: on Tauri v2,
```diff
- const isTauri = '__TAURI__' in window;
+ const isTauri = '__TAURI_INTERNALS__' in window;
```

**Reply by ctsstc** [reply]
I wonder if `isMobile` would turn on for a desktop/laptop with a touch screen? Or does the desktop version not get that support I wonder? Or is that what makes something "feel mobile" although I'm sure that this is meant to detect the build-type. Maybe if you're trying to detect the feature of maxTouchPoints then this would be more applicable, but not aptly named.

**Comment by ilyagr** [comment]
I found an `isTauri` function in https://github.com/tauri-apps/tauri/blob/e349dfe5722c0b47620582bd03f8c184ffeb6979/packages/api/src/core.ts#L317 that seems to work, but doesn't seem to be documented. Is it "officially supported" and/or supposed to be more future-proof than checking for `__TAURI_INTERNALS__`?

**Update:** It seems to be legit. It was added in 2.0.0-beta.9, in #9539.

**Comment by loganpowell** [comment]
Just an update that this is now `window.__TAURI_INTERNALS__`

**Reply by loganpowell** [reply]
`window.isTauri` also works

**Reply by justingolden21** [reply]
Perfect, thank you! Was migrating my tauri app from 1 to 2 and couldn't figure out why my func to check if it's tauri wasn't working.

In case this helps someone:

`return typeof window !== 'undefined' && (window as any).__TAURI_INTERNALS__;`

**Comment by vmiguellima** [comment]
Tauri already provides this through:

```rs
import { isTauri } from "@tauri-apps/api/core";

isTauri() ? doThis : doThat;
```

**Comment by wolffan** [comment]
If I'm compiling the app locally (Mac), should `isTauri` be available?
I get false when running dev env

**Comment by buraktabn** [comment]
What about checking if isTauri in the server side? `window` will be undefined there

**Comment by andersmmg** [comment]
Here's what I'm using, a modified version of the several options provided above. Works with the latest Tauri, and provides doc comments.
```ts
/** Running on a desktop app or a mobile app, but not in the browser */
const isTauri = '__TAURI_INTERNALS__' in window;

/** Running in the browser on either desktop or mobile, but not as a tauri app */
export const isWeb = !isTauri;

/** Running on mobile either in the browser or as a tauri app */
export const isMobile = navigator.maxTouchPoints > 0;
/** Running on desktop either in the browser or as a tauri app */
export const isDesktop = !isMobile;

/** Running on mobile as a tauri app, but not on the browser */
export const isTauriMobile = isTauri && isMobile;
/** Running on desktop as a tauri app, but not on the browser */
export const isTauriDesktop = isTauri && isDesktop;

/** Running on mobile in the browser, but not as a tauri app */
export const isWebMobile = isWeb && isMobile;
/** Running on desktop in the browser, but not as a tauri app */
export const isWebDesktop = isWeb && isDesktop;

export default isTauri;
```

**Comment by Shyam-Chen** [comment]
```ts
import { isTauri } from '@tauri-apps/api/core';
import { platform } from '@tauri-apps/plugin-os';

if (isTauri()) {
  console.log('platform =', platform()); // 'linux' | 'windows' | 'macos' | 'ios' | 'android'
} else {
  console.log('platform = web');
}
```

```ts
import { isTauri } from '@tauri-apps/api/core';
import { platform } from '@tauri-apps/plugin-os';
import mobile from 'is-mobile';

export const IS_TAURI = isTauri();
export const IS_WEB = !IS_TAURI;

const tauriPlatform = IS_TAURI ? platform() : null;

export const IS_MOBILE = IS_TAURI
  ? tauriPlatform === 'ios' || tauriPlatform === 'android'
  : mobile();

export const IS_DESKTOP = !IS_MOBILE;

export const IS_TAURI_MOBILE = IS_TAURI && IS_MOBILE;
export const IS_TAURI_DESKTOP = IS_TAURI && IS_DESKTOP;
export const IS_WEB_MOBILE = IS_WEB && IS_MOBILE;
export const IS_WEB_DESKTOP = IS_WEB && IS_DESKTOP;
```
